src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/BuildAssetsUtils.cs

```diff
@@ -0,0 +1,256 @@
+        /// A single framework will always return true.
+        /// </summary>
+        private static bool NeedsMSBuildConditionals(Dictionary<NuGetFramework, TargetsAndProps> buildAssetsByFramework)
+        {
+            if (buildAssetsByFramework.Count > 1)
+            {
+                var combinedAssets = buildAssetsByFramework.Select(entry =>
+                    entry.Value.Targets
+                        .Concat(entry.Value.Props)
+                        .OrderBy(s => s, StringComparer.Ordinal)
+                        .ToArray())
+                   .ToArray();
+
+                // Compare all groups to the first group
+                return combinedAssets.Skip(1).Any(group => !combinedAssets[0].SequenceEqual(group));
+            }
+
+            return false;
+        }
+
+        private static TargetsAndProps GetTargetsAndPropsForFramework(
+            IEnumerable<RestoreTargetGraph> targetGraphs,
+            IReadOnlyList<NuGetv3LocalRepository> repositories,
+            RemoteWalkContext context,
+            RestoreRequest request,
+            Dictionary<RestoreTargetGraph,
+            Dictionary<string, LibraryIncludeFlags>> includeFlagGraphs,
+            NuGetFramework projectFramework)
+        {
+            var result = new TargetsAndProps();
+
+            // Skip runtime graphs, msbuild targets may not come from RID specific packages
+            var graph = targetGraphs
+                .Single(g => string.IsNullOrEmpty(g.RuntimeIdentifier) && g.Framework.Equals(projectFramework));
+
+            // Gather props and targets to write out
+            var buildGroupSets = GetMSBuildAssets(context, graph, request.Project, includeFlagGraphs);
+
+            // Second find the nearest group for each framework
+            foreach (var buildGroupSetsEntry in buildGroupSets)
+            {
+                var libraryIdentity = buildGroupSetsEntry.Key;
+                var buildGroupSet = buildGroupSetsEntry.Value;
+
+                // Find the nearest msbuild group, this can include the root level Any group.
+                var buildItems = NuGetFrameworkUtility.GetNearest(
+                        buildGroupSet,
+                        graph.Framework,
+                        group =>
+                            group.Properties[ManagedCodeConventions.PropertyNames.TargetFrameworkMoniker]
+                                as NuGetFramework);
+
+                // Check if compatible build assets exist
+                if (buildItems != null)
+                {
+                    AddPropsAndTargets(repositories, libraryIdentity, buildItems, result);
+                }
+            }
+
+            return result;
+        }
+
+        /// <summary>
+        /// Check if NUGET_XPROJ_WRITE_TARGETS is true.
+        /// </summary>
+        private static bool ForceWriteTargets()
+        {
+            var envVar = Environment.GetEnvironmentVariable("NUGET_XPROJ_WRITE_TARGETS");
```

Comment:
ok..my thinking is that if the string has any special significance, we should declare it as a const...but use your discretion here.

---
In reply to: [71958613](https://github.com/NuGet/NuGet.Client/pull/782#discussion_r71958613) [](ancestors = 71958613)